Add edit and view changes features

mikeebowen opened the pull request:
## This pull request adds 2 new features: 

- User can edit xml parts in VS Code and those changes will be reflected in the OOXML document. E.G. if a user edits a .docx file in VS Code the changes will be reflected when the document is opened in Word or another document editor.
- After editing from an outside program, xml parts that have been changed will be reflected with a yellow asterisk for edited, a green asterisk for created, and a red asterisk for deleted. If a file has been changed the user can right click for the context menu and click "Compare with Previous" to view a diff of the previous version of the xml part and the current version.
- Tests have been added for the new features.
- Documentation in the README has been updated to show users how to user the new features.

yuenm18 commented on .eslintrc.js:
```diff
@@ -0,0 +1,37 @@
+      },
+    ],
+    'func-call-spacing': ['error', 'never'],
+    indent: [2, 2, { SwitchCase: 1 }],
```
```suggestion
    indent: ['error', 2, { SwitchCase: 1 }],
```

yuenm18 commented on .eslintrc.js:
```diff
@@ -0,0 +1,37 @@
+    ],
+    'no-console': ['warn', { allow: ['warn', 'error'] }],
+    'no-unused-vars': 'off',
+    quotes: [2, 'single', { avoidEscape: true }],
```
```suggestion
    quotes: ['error', 'single', { avoidEscape: true }],
```

yuenm18 commented on src/ooxml-viewer.ts:
```diff
@@ -1,99 +1,600 @@
+          const folderPath = join(this.fileCachePath, dirname(fileNode.fullPath));
+          const filePath: string = join(folderPath, fileNode.fileName);
+          await this._createFile(fileNode.fullPath, fileNode.fileName, true);
+          const uri: Uri = Uri.parse(`file:///${filePath}`);
```
```suggestion
          const uri: Uri = Uri.file(filePath);
```

yuenm18 commented on src/ooxml-viewer.ts:
```diff
@@ -1,99 +1,600 @@
+      const enc = new TextEncoder();
+      const rightUri = Uri.file(filePath);
+      const leftUri = Uri.file(compareFilePath);
+      const rightXml = await (await workspace.fs.readFile(rightUri)).toString();
```
```suggestion
      const rightXml = (await workspace.fs.readFile(rightUri)).toString();
```

yuenm18 commented on src/ooxml-viewer.ts:
```diff
@@ -1,99 +1,600 @@
+      const rightUri = Uri.file(filePath);
+      const leftUri = Uri.file(compareFilePath);
+      const rightXml = await (await workspace.fs.readFile(rightUri)).toString();
+      const leftXml = await (await workspace.fs.readFile(leftUri)).toString();
```
```suggestion
      const leftXml = (await workspace.fs.readFile(leftUri)).toString();
```

yuenm18 commented on src/ooxml-viewer.ts:
```diff
@@ -1,99 +1,600 @@
+  }
+  /**
+   * Closes all active editor tabs
+   * @method _closeEditors
```
```suggestion
   * @method closeEditors
```

yuenm18 commented on src/ooxml-viewer.ts:
```diff
@@ -1,99 +1,600 @@
+      const filePath: string = join(folderPath, fileName);
+      await workspace.fs.createDirectory(Uri.file(folderPath));
+      const file: JSZipObject | null = this.zip.file(relativePath);
+      const text: string = (await file?.async('text')) ?? (await (await workspace.fs.readFile(Uri.file(preFilePath))).toString());
```
```suggestion
      const text: string = (await file?.async('text')) ?? (await workspace.fs.readFile(Uri.file(preFilePath))).toString();
```

yuenm18 commented:
Thank you for the PR, it looks like these changes will be very useful

I have a couple of general comments:
- I noticed an issue when you click a folder in the "OOXML Viewer." It opens up the first file in the folder rather than just opening the dropdown
- Lets make the next version 1.0.0 and have the changelog file contain only the original entry for version 0.0.1 and a new entry for version 1.0.0 with all your changes since version 0.0.1 because the extension only has version 0.0.1 published
- Lets get rid of the releases folder, I want to tag the commit that the version was built off of and add the vsix file to corresponding the github release rather than committing the visx files
- Can you add an extra newlines between your methods

mikeebowen commented:
Thanks for the suggestions, obviously .toString() is not asynch 🤦‍♂️. 
- I will look into the folder displaying a file. 
- I can remove the tags and revert to v1.0.0 
- I need to keep the releases folder for now, because some coworkers are dog fooding the extension to help me find bugs and that is how I give them the releases. I don't have any objection to you deleting it after the pull request is approved before publishing, though. 
- I can add extra new lines.
- I used a cli tool called [auto-changelog](https://github.com/cookpete/auto-changelog) to create the changelog. It works really well, but because I ran it on my fork it lists all the commits as from my fork. After the pull request is approved, you should rerun auto-changelog so the changelog has the correct links to the commits in your repo.
I will try and have an updated pull request for you tomorrow or the next day. I'll keep you posted.

mikeebowen commented:
OK, the remaining issues have been taken care of
- Stopped calling viewFile when the user clicks on a FileNode that is a folder
- Reverted the versions and it now goes from v0.0.1 to v1.0.0 and the CHANGELOG.md has been updated with auto-changelog
- Deleted the releases folder, I can distribute the vsix other ways
- Added lines between the methods

Hopefully that takes care of everything 🙂
